### PR TITLE
Decoy Balloons

### PR DESCRIPTION
**Pull request by Shadowmech88**
Traitors can now buy a decoy balloon for 1 TC. When inflated, it copies your current appearance and your name.
That's it.

:cl:
 * rscadd: Added decoy balloons to traitor uplinks, with a price of 1 TC each.

### CODE REVIEW

File: code/datums/uplink_item.dm [hunk 1]
```diff
@@ -477,12 +477,18 @@ var/list/uplink_items = list()
 	item = /obj/item/device/chameleon
 	cost = 6
 
-/datum/uplink_item/stealthy_tools/chameleon_proj
+/datum/uplink_item/stealthy_tools/smoke_bombs
```
Comment (Probe1) [inline code/datums/uplink_item.dm:480]: wait what

**Comment by Ephx1** [comment]
Make it so it can say one small line of player-input speech too.
_You're pretty good_

File: code/datums/uplink_item.dm (hunk 1 above)
Comment (Shadowmech88) [reply code/datums/uplink_item.dm:480]: The path was the same as the chameleon projector's.
If the atomicity is *really* a problem I could remove it, but it was just a one-line fix.

**Comment by Blithering** [comment]
Good luck finding me, I'm behind 20 decoys!

**Comment by SkowronX** [comment]
Seconding Ephx1's idea.

File: code/datums/uplink_item.dm (hunk 1 above)
Comment (9600bauds) [reply code/datums/uplink_item.dm:480]: Whoops

**Comment by jknpj** [comment]
dummy suits

Decoy launcher when?

**Comment by Shadowmech88** [comment]
I feel like making it repeat a specific phrase would make it even less effective than having it stand in place silently.

**Comment by calzilla1** [comment]
Make it so if they hit it with a blunt object or fists they have a chance to get knocked down

File: code/game/objects/items/toys.dm [hunk 2]
```diff
@@ -1292,4 +1293,35 @@
...
+	inflated_type = /obj/item/toy/balloon/inflated/decoy
+	volume = 120	//liters
+
+/obj/item/toy/balloon/decoy/inflate(mob/user, datum/gas_mixture/G)
+	var/obj/item/toy/balloon/inflated/decoy/D = ..()
+	if(!istype(D))
+		return
+	user.drop_item(D, force_drop = 1)
+	D.name = user.name
+	D.desc = user.desc
```
Comment (PJB3005) [inline code/game/objects/items/toys.dm]: Should probably copy over the examine text, somehow.

**Review by PJB3005 — APPROVED** [review]

**Comment by Probe1** [comment]
Give them the code from punching bags so they wiggle when you punch them.

**Comment by Ephx1** [comment]
Just have it say the line when you interact with it.

File: code/game/objects/items/toys.dm [hunk 3]
```diff
@@ -1292,4 +1293,35 @@
...
+	col = null
+	inflated_type = /obj/item/toy/balloon/inflated/decoy
+	volume = 120	//liters
+
+/obj/item/toy/balloon/decoy/inflate(mob/user, datum/gas_mixture/G)
+	var/obj/item/toy/balloon/inflated/decoy/D = ..()
+	if(!istype(D))
+		return
+	user.drop_item(D, force_drop = 1)
+	D.name = user.name
```
Comment (Exxion) [inline code/game/objects/items/toys.dm]: `name` is part of `appearance`, so explicitly copying it isn't necessary

File: code/game/objects/items/toys.dm (hunk 2 above)
Comment (Exxion) [reply code/game/objects/items/toys.dm]: Like
All of it?
Oh dear fuck

Comment (PJB3005) [reply code/game/objects/items/toys.dm]: Of course.

I don't mean copy over the code, just copy the text outputted itself.

**Comment by Shadowmech88** [comment]
* An inflatable decoy's examine text now exactly matches that of the person who inflated it, rendering them indistinguishable at first glance.

**Review by PJB3005 — APPROVED** [review]

File: code/game/objects/items/toys.dm [hunk 4]
```diff
@@ -1292,4 +1293,40 @@
...
+	inflated_type = /obj/item/toy/balloon/inflated/decoy
+	volume = 120	//liters
+
+/obj/item/toy/balloon/decoy/inflate(mob/user, datum/gas_mixture/G)
+	var/obj/item/toy/balloon/inflated/decoy/D = ..()
+	if(!istype(D))
+		return
+	user.drop_item(D, force_drop = 1)
+	D.appearance = user.appearance
+	var/datum/log/L = new
```
Comment (Exxion) [inline code/game/objects/items/toys.dm:1327]: Is this datum ever deleted?

Comment (Shadowmech88) [reply code/game/objects/items/toys.dm:1327]: No, good catch.

Comment (PJB3005) [reply code/game/objects/items/toys.dm:1327]: It would automatically be caught by BYOND's reference counter anyways.

**Comment by Shadowmech88** [comment]
How does this look?
http://puu.sh/tDmHA/0073dc366a.webm

**Comment by Exxion** [comment]
Damn excellent
That's how it looks

**Comment by Shadowmech88** [comment]
* A decoy balloon will now wobble when hit. See: http://puu.sh/tDmHA/0073dc366a.webm
* While deflated, a phrase for the decoy balloon to say may be set using a verb. Once inflated, the decoy balloon will say this phrase whenever it is hit.

File: code/game/objects/items/toys.dm (hunk 4 above)
Comment (Shadowmech88) [reply code/game/objects/items/toys.dm:1327]: I added a `qdel(L)` anyway, just to be safe.

**Comment by Lokiamis** [comment]
I think it would be good if tasers and flashbangs popped them

**Comment by Ephx1** [comment]
Rad

**Comment by Probe1** [comment]
Yeah I agree with @Lokiamis

File: code/game/objects/items/toys.dm [hunk 5]
```diff
@@ -1292,4 +1293,72 @@
...
+	var/list/hit_sounds = list('sound/weapons/genhit1.ogg', 'sound/weapons/genhit2.ogg', 'sound/weapons/genhit3.ogg',\
+	'sound/weapons/punch1.ogg', 'sound/weapons/punch2.ogg', 'sound/weapons/punch3.ogg', 'sound/weapons/punch4.ogg')
+
+/obj/item/toy/balloon/inflated/decoy/examine(mob/user, var/size = "")
+	if(desc)
+		to_chat(user, desc)
+
+/obj/item/toy/balloon/inflated/attackby(obj/item/weapon/W, mob/user)
+	..()
+	if(!src.gcDestroyed)
```
Comment (PJB3005) [inline code/game/objects/items/toys.dm:1350]: This seems oddly specific.

Comment (Shadowmech88) [reply code/game/objects/items/toys.dm:1350]: I need to verify that it wasn't popped in the parent call before assuming it was hit with a blunt object and calling `attack_hand()`.

Comment (PJB3005) [reply code/game/objects/items/toys.dm:1350]: Yeah but the qdel should make it impossible for anybody to be able to hit the thing anyways so?

Comment (Exxion) [reply code/game/objects/items/toys.dm:1350]: It could be popped and bopped *at the same time*

Comment (PJB3005) [reply code/game/objects/items/toys.dm:1350]: Yes but as soon as it gets qdel'd it moves to nullspace and it becomes impossible for it to still be bopped.

Comment (Shadowmech88) [reply code/game/objects/items/toys.dm:1350]: Being hit by an object causes the bopping, wobbling, and quote saying just like if you were to punch it.
However, I need that to not happen if the object hitting it causes it to pop. This means I need a way to check whether it's been popped after the `..()` call. Is there a better way to do that than checking `gcDestroyed`?

Comment (PJB3005) [reply code/game/objects/items/toys.dm:1350]: Oh, ok then.

**Comment by Shadowmech88** [comment]
It makes sense for projectiles to pop it, but why would a flashbang?

**Comment by Shadowmech88** [comment]
Also projectiles *already* pop balloons, of all types.

**Comment by Probe1** [comment]
No no lets get this show on the road.

**Pull request merged**